linera-core/src/local_node.rs

```diff
@@ -0,0 +1,546 @@
+        application_id: UserApplicationId,
+    ) -> Result<UserApplicationDescription, LocalNodeError> {
+        let mut node = self.node.lock().await;
+        let response = node
+            .state
+            .describe_application(chain_id, application_id)
+            .await?;
+        Ok(response)
+    }
+
+    pub async fn download_certificates<A>(
+        &mut self,
+        mut validators: Vec<(ValidatorName, A)>,
+        chain_id: ChainId,
+        target_next_block_height: BlockHeight,
+    ) -> Result<ChainInfo, LocalNodeError>
+    where
+        A: ValidatorNode + Send + Sync + 'static + Clone,
+    {
+        // Sequentially try each validator in random order.
+        validators.shuffle(&mut rand::thread_rng());
+        for (name, client) in validators {
+            let info = self.local_chain_info(chain_id).await?;
+            if target_next_block_height <= info.next_block_height {
+                return Ok(info);
+            }
+            self.try_download_certificates_from(
+                name,
+                client,
+                chain_id,
+                info.next_block_height,
+                target_next_block_height,
+            )
+            .await?;
+        }
+        let info = self.local_chain_info(chain_id).await?;
+        if target_next_block_height <= info.next_block_height {
+            Ok(info)
+        } else {
+            Err(LocalNodeError::CannotDownloadCertificates {
+                chain_id,
+                target_next_block_height,
+            })
+        }
+    }
+
+    /// Downloads and stores the specified blobs, unless they are already in the cache or storage.
+    ///
+    /// Does not fail if a blob can't be downloaded; it just gets omitted from the result.
+    pub async fn read_or_download_blobs<A>(
+        &mut self,
+        validators: Vec<(ValidatorName, A)>,
+        blob_locations: impl IntoIterator<Item = (BytecodeLocation, ChainId)>,
+    ) -> Result<Vec<HashedValue>, LocalNodeError>
+    where
+        A: ValidatorNode + Send + Sync + 'static + Clone,
+    {
+        let mut blobs = vec![];
+        let mut tasks = vec![];
+        let mut node = self.node.lock().await;
+        for (location, chain_id) in blob_locations {
+            if let Some(blob) = node.state.recent_value(&location.certificate_hash).await {
+                blobs.push(blob);
+            } else {
+                let validators = validators.clone();
+                let storage = node.state.storage_client().clone();
+                tasks.push(Self::read_or_download_blob(
+                    storage, validators, chain_id, location,
+                ));
+            }
+        }
+        drop(node); // Free the lock while awaiting the tasks.
+        if tasks.is_empty() {
+            return Ok(blobs);
+        }
+        let results = future::join_all(tasks).await;
+        let mut node = self.node.lock().await;
+        for result in results {
+            if let Some(blob) = result? {
```

Comment:
Wait so it does fail if the blob cannot be downloaded.

Reply:
@afck

Reply:
`read_or_download_blob` only returns an error if there was a _storage_ error. If the blob couldn't be downloaded it returns `Ok(None)`.